linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala

```diff
@@ -0,0 +1,1496 @@
+    private val TypeErrorRef = globalRef("TypeError")
+
+    // Conditional global references that we often use
+    private def BigIntRef = globalRef("BigInt")
+
+    private val classData = Ident("$classData")
+
+    private val orderedPrimitiveClasses = {
+      List(VoidClass, BooleanClass, CharClass, ByteClass, ShortClass, IntClass,
+          LongClass, FloatClass, DoubleClass)
+    }
+
+    def build(): WithGlobals[Tree] = {
+      defineLinkingInfo()
+      defineJSBuiltinsSnapshotsAndPolyfills()
+      declareCachedL0()
+      definePropertyName()
+      defineCharClass()
+      defineRuntimeFunctions()
+      defineDispatchFunctions()
+      defineDoubleToIntLongConversions()
+      defineES2015LikeHelpers()
+      defineModuleHelpers()
+      defineIntrinsics()
+      defineIsHijackedBoxedClasses()
+      defineAsHijackedBoxedClasses()
+      defineBoxFunctions()
+      defineTypedArrayConversions()
+      defineTypeDataClass()
+      defineIsArrayOfPrimitiveFunctions()
+      defineAsArrayOfPrimitiveFunctions()
+      definePrimitiveTypeDatas()
+
+      WithGlobals(Block(buf.result()), trackedGlobalRefs)
+    }
+
+    private def defineLinkingInfo(): Unit = {
+      // must be in sync with scala.scalajs.runtime.LinkingInfo
+
+      def checkedBehaviorInt(behavior: CheckedBehavior): IntLiteral = {
+        int(behavior match {
+          case CheckedBehavior.Compliant => 0
+          case CheckedBehavior.Fatal     => 1
+          case CheckedBehavior.Unchecked => 2
+        })
+      }
+
+      val linkingInfo = ObjectConstr(List(
+          str("semantics") -> ObjectConstr(List(
+              str("asInstanceOfs") -> checkedBehaviorInt(asInstanceOfs),
+              str("arrayIndexOutOfBounds") -> checkedBehaviorInt(arrayIndexOutOfBounds),
+              str("moduleInit") -> checkedBehaviorInt(moduleInit),
+              str("strictFloats") -> bool(strictFloats),
+              str("productionMode") -> bool(productionMode)
+          )),
+          str("assumingES6") -> bool(useECMAScript2015),
+          str("linkerVersion") -> str(ScalaJSVersions.current),
+          str("globalThis") -> This()
+      ))
+
+      val linkingInfoVarRef = envField("linkingInfo")
+
+      buf ++= List(
+          const(linkingInfoVarRef, linkingInfo),
+          Apply(genIdentBracketSelect(ObjectRef, "freeze"), linkingInfoVarRef :: Nil),
+          Apply(genIdentBracketSelect(ObjectRef, "freeze"),
+              BracketSelect(linkingInfoVarRef, str("semantics")) :: Nil)
+      )
+    }
+
+    private def defineJSBuiltinsSnapshotsAndPolyfills(): Unit = {
+      def genPolyfillFor(builtinName: String): Tree = builtinName match {
+        case "imul" =>
+          val a = varRef("a")
+          val b = varRef("b")
+          val ah = varRef("ah")
+          val al = varRef("al")
+          val bh = varRef("bh")
+          val bl = varRef("bl")
+          Function(arrow = false, paramList(a, b), Block(
+              const(ah, a >>> 16),
```

Comment:
Yes that's right. Actually even if the operand is not 32 bits it will be coerced to 32 bits before doing the shift. The `& 0xffff` was completely redundant.